crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -43,154 +42,464 @@ mod validate_op_tests;
 ) -> WorkflowResult<WorkComplete> {
-    let complete =
-        sys_validation_workflow_inner(workspace, space, network, sys_validation_trigger).await?;
+    // Run the actual sys validation using data we have locally
+    let outcome_summary =
+        sys_validation_workflow_inner(workspace.clone(), current_validation_dependencies.clone())
+            .await?;
+
+    // trigger app validation to process any ops that have been processed so far
+    if outcome_summary.accepted > 0 {
+        tracing::debug!("Sys validation accepted {} ops", outcome_summary.accepted);
+
+        trigger_app_validation.trigger(&"sys_validation_workflow");
+    }
+
+    // Now go to the network to try to fetch missing dependencies
+    let network_cascade = Arc::new(workspace.network_and_cache_cascade(network));
+    let missing_action_hashes = current_validation_dependencies.lock().get_missing_hashes();
+    let num_fetched: usize = futures::stream::iter(missing_action_hashes.into_iter().map(|hash| {
+        let network_cascade = network_cascade.clone();
+        let current_validation_dependencies = current_validation_dependencies.clone();
+        async move {
+            match network_cascade.retrieve(hash, Default::default()).await {
+                Ok(Some((record, source))) => {
+                    let mut deps = current_validation_dependencies.lock();
+
+                    // If the source was local then that means some other fetch has put this action into the cache,
+                    // that's fine we'll just grab it here.
+                    if deps.insert(record, source) {
+                        1
+                    } else {
+                        0
+                    }
+                }
+                Ok(None) => {
+                    // This is fine, we didn't find it on the network, so we'll have to try again.
+                    // TODO This will hit the network again fairly quickly if sys validation is triggered again soon.
+                    //      It would be more efficient to wait a bit before trying again.
+                    0
+                }
+                Err(e) => {
+                    tracing::error!(error = ?e, "Error fetching missing dependency");
+                    0
+                }
+            }
+        }
+        .boxed()
+    }))
+    .buffer_unordered(10)
+    .collect::<Vec<usize>>()
+    .await
+    .into_iter()
+    .sum();
+
+    if num_fetched > 0 {
+        // If we fetched anything then we can re-run sys validation
+        trigger_self.trigger(&"sys_validation_workflow");
+    }
 
-    // --- END OF WORKFLOW, BEGIN FINISHER BOILERPLATE ---
+    // Finally, for anything we've fetched from the network we need to push it into the incoming dht ops workflow
+    let network_fetched_action_hashes = current_validation_dependencies
+        .lock()
+        .get_network_fetched_hashes();
+    futures::future::join_all(network_fetched_action_hashes.into_iter().map(|hash| {
```

Comment:
This is really a lot of extra work, and tracking enough state to be able to reconstruct DHT ops is adding complexity to this entire workflow. This wasn't as obvious before as we were just doing an extra `retrieve` against the cascade. That was inefficient though. I think we should refactor this trigger so we just pass a list of action hashes to the incoming_ops workflow and it can figure out for itself which DHT ops it needs to process based on that. That idea probably needs some refining, but something along those lines!

Reply:
Happily this is gone! Which also means that use of `retrieve` has been replaced with `retrieve_action` because we don't need entries for previous DhtOps any more and all the mess related to holding onto that data has been stripped out. Very happy with how much simpler this makes the workflow :)